processing/src/main/java/io/druid/segment/data/GenericIndexed.java

```diff
@@ -109,6 +269,83 @@
     return new GenericIndexed<T>(theBuffer.asReadOnlyBuffer(), strategy, allowReverseLookup);
   }
 
+  public static <T> GenericIndexed<T> read(ByteBuffer buffer, ObjectStrategy<T> strategy)
+  {
+    byte versionFromBuffer = buffer.get();
+
+    if (version_one == versionFromBuffer) {
+      return createVersionOneGenericIndexed(buffer, strategy);
+    } else if (version_three == versionFromBuffer) {
+      throw new IAE(
+          "use read(ByteBuffer buffer, ObjectStrategy<T> strategy, SmooshedFileMapper fileMapper) to read version 3 indexed.",
+          versionFromBuffer
+      );
+    }
+    throw new IAE("Unknown version[%s]", versionFromBuffer);
+  }
```

Comment:
Other read(with smoosher) will again read version from the buffer, in that case it will error out.